core/src/main/java/io/grpc/internal/ServiceConfigState.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import io.grpc.Status;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ServiceConfigState} holds the state of the current service config.  It must be mutated
+ * and read from {@link ManagedChannelImpl} constructor or the provided syncContext.
+ */
+final class ServiceConfigState {
+  @Nullable private final ManagedChannelServiceConfig defaultServiceConfig;
+  private final boolean lookUpServiceConfig;
+
+  // mutable state
+  @Nullable private ManagedChannelServiceConfig currentServiceConfig;
+  @Nullable private Status currentError;
+  // Has there been at least one successful update?
+  private boolean updated;
+
+  /**
+   * @param defaultServiceConfig The initial service config, or {@code null} if absent.
+   * @param lookUpServiceConfig {@code true} if service config updates might occur.
+   * @param syncCtx The synchronization context that this is accessed from.
+   */
+  ServiceConfigState(
+      @Nullable ManagedChannelServiceConfig defaultServiceConfig,
+      boolean lookUpServiceConfig) {
+    this.defaultServiceConfig = defaultServiceConfig;
+    this.lookUpServiceConfig = lookUpServiceConfig;
+    if (!lookUpServiceConfig) {
+      this.currentServiceConfig = defaultServiceConfig;
+    }
+  }
+
+  /**
+   * Returns {@code true} if it RPCs should wait on a service config resolution.  This can return
+   * {@code false} if:
+   *
+   * <ul>
+   *   <li>There is a valid service config from the name resolver
+   *   <li>There is a valid default service config and a service config error from the name
+   *       resolver
+   *   <li>No service config from the name resolver, and no intent to lookup a service config.
+   * </ul>
+   *
+   * <p>In the final case, the default service config may be present or absent, and will be the
+   * current service config.
+   */
+  boolean shouldWaitOnServiceConfig() {
+    return !(updated || !expectUpdates());
+  }
+
+  /**
+   * Gets the current service config.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable ManagedChannelServiceConfig getCurrentServiceConfig() {
+    checkState(!shouldWaitOnServiceConfig(), "still waiting on service config");
+    return currentServiceConfig;
+  }
+
+  /**
+   * Gets the current service config error.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable Status getCurrentError() {
```

Comment:
I wonder if the caller would care about the error only when currentServiceConfig is null. In other words, I wonder if it would be sufficient to have a `ConfigOrError` field rather than `currentServiceConfig` and `currentError`.

Reply:
ConfigOrError involves unnecessary casting in MCI.

Reply:
Also, right now, ConfigOrError doesn't allow null, so then the ConfigOrError field would have to be nullable.   Leaving it as a MCSC simplifies MCI's usage of  getCurrentServiceConfig()

Reply:
It's natural for a `ConfigOrError` field to be nullable, because it's the case in `ResolutionResult`. On the other hand, the config part of `ConfigOrError` is not nullable in the no-error case, while you keep it as a nullable field here, which I think is a mismatch and that involves conversion.

If you made both `update()` and `get()` deal with `ConfigOrError`, that would simplify the interface and bring better consistency.

Casting is unfortunate, but it's already the case on the API so it's not a new issue.